Add support for Bitbucket 2FA via OAuth

mminns opened the pull request:
Hi
This PR adds formal support for Bitbucket to the GCM and allows the use of Bitbucket 2FA by way of OAuth tokens.

I have attempted to keep the changes as isolated as possible within the Bitbucket DLLs/Projects, however there are a few changes to shared resources. 
I have attempted to follow the pattern provided by the GitHub support, but again I didn't want to disrupt existing code, so rather then just copy into the Bitbucket DLL I've placed code that could/should be shared into the Core.Authentication DLL/Project. It could potentially be reused by other providers.


Support for Basic Auth Bitbucket credentials is almost identical to the existing GitHub approach. The real changes are around 2FA/OAuth support in _BitbucketAuthority.AquireToken()_ .
This relies on the fact that Bitbucket's REST API will respond with a 403 if you provide a valid username/password but 2FA has been enabled.

When this is detected the user is shown a second dialog with the ability to trigger an OAuth dance.
Assuming a successful dance the OAuth access token is retrieved by running a local HTTP endpoint, the _SimpleServer_ class, which is used as a callback URL.

Bitbucket OAuth tokens are short-lived, 1 hour, but Bitbucket provides an automated process using valid refresh tokens to request new OAuth tokens if required. As such the Bitbucket implementation will store both the access_token and the refresh_token in the Windows Credentials Manager/Vault.

When stored access_tokens are retrieved from the Vault they are tested by running a REST request, if that fails an attempt is made to retrieve a new access_token using the stored refresh_token, if that succeeds the new access_token is stored before returning it to git.


Other than that I am aware of a couple of issues:
1) It could do with more automated testing
2) In use it appears to be more likely to delete credentials and re-prompt the user for new ones that I really expect, but it is possible that is because I run a dev environment with multilple git installs etc and I'm trigger other instances of the GCM that don't recognize Bitbucket 2FA and are then deleting the credentials.

Finally as I'm sure I've trampled on a number of conventions in this work, so please call me out on anything you would like explained further and areas in need of rework :)

Thanks

msftclas commented:
@mminns,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

whoisj commented:
Wow, Mr. Minns you've been very busy. :astonished: This is quite exciting - thank you :smiley: 🙇 

As you can imagine, it'll take me a bit to get this reviewed. Let me sit down my time lords and figure out the best way to tackle this while still getting VS 2017 finished. :sweat_smile:

mminns commented:
Ha ha ha

No problem. To be honest I just  want to get the process started... I have
no assumptions about when it may end :)

Mike

On 1 March 2017 at 21:44, J Wyman <notifications@github.com> wrote:

> Wow, Mr. Minns you've been very busy. 😲 This is quite exciting - thank
> you 😃 🙇
>
> As you can imagine, it'll take me a bit to get this reviewed. Let me sit
> down my time lords and figure out the best way to tackle this while still
> getting VS 2017 finished. 😅
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/Git-Credential-Manager-for-Windows/pull/372#issuecomment-283480311>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAxwA0ElWT2pjAhxFvv1URaS_BUxO5MSks5rhebIgaJpZM4MQNsf>
> .
>



-- 
*Mike Minns*
mminns@atlassian.com
SourceTree Development Team Lead
https://www.sourcetreeapp.com
https://www.atlassian.com

whoisj commented:
@minns I spend the better part of the day re-working your branch to be compatible with the latest master. Please take a look here: https://github.com/whoisj/Git-Credential-Manager-for-Windows/tree/bitbucket-rebased for reference.

There was a fair amount of rework required, and I didn't take all of your commits verbatim into my branch (I squashed a lot of them). By no means am I expecting you to take my branch as is. Feel free to re-work your own commits as you see fit, but you'll need to align with something similar to what I've pushed for you.

Generally, things are looking good. I think my biggest concerns are:

 1. I have no way to functionally test this. I do not have a Bitbucket account - I'll need to trust that you've tested the functionality and that it actually works.
 2. There are no attribution, copyright, or licensing headers at the top of any of your files. You'll at least want to add something to the `BITBUCKET_README.md` you've added. I think we can all fairly assume this isn't just a public domain contribution.
 3. There are still quite a few "TODO" statements lying about. Does this mean the code is not ready and should not be accepted into GCM master?

Thanks - again, this was awesome - I really look forward to seeing it lands. 😄

mminns commented:
Fantastic, I hope the re-work wasn't too arduous! 
Its likely to be next week before I get a real chance to look into it, but to answer one question I'm using it daily at the moment, having happily dropped SSH access to Bitbucket. I believe it is complete but obviously in need of some tidying up.

Thanks again

ps. private Bitbucket accounts are free ;)

whoisj commented:
> ps. private Bitbucket accounts are free ;)

Well, I'll be darned - did not know that (you really should give your marketing people some flak :wink:). I'll set one up to help validate things then - thanks.

mminns commented:
OK, here we go after some delay mk2.

whoisj commented:
❗️ 😀 ❗️ I'll review this ASAP!

whoisj commented on Bitbucket.Authentication/Controls/Text/MaskedPasswordBox.cs:
```diff
@@ -0,0 +1,113 @@
+                        Debug.Assert(password != null, "Password can't be null here");
+                        // Replace or insert char
+                        string currentCharacter = currentText[i].ToString(CultureInfo.InvariantCulture);
+                        password = BaseText.Length == password.Length ? password.Remove(i, 1).Insert(i, currentCharacter) : password.Insert(i, currentCharacter);
```
is this safe? are we certain `i` is always within the bounds of `[0, password.length)`?

mminns replied:
I'll investigate.

FWIW this is a straight copy of https://github.com/Microsoft/Git-Credential-Manager-for-Windows/blob/master/GitHub.Authentication/Controls/Text/MaskedPasswordBox.cs

As such maybe it should move to GUI-Shared anyway?

whoisj replied:
> As such maybe it should move to GUI-Shared anyway?

I'm OK with that, but @Haacked is the right person to ask about moving code into or out of the GitHub sub-project. Phil?

whoisj commented on Bitbucket.Authentication/Controls/Text/MaskedPasswordBox.cs:
```diff
@@ -0,0 +1,113 @@
+                        password = BaseText.Length == password.Length ? password.Remove(i, 1).Insert(i, currentCharacter) : password.Insert(i, currentCharacter);
+                    }
+                }
+                Debug.Assert(password != null, "Password can't be null here");
```
is this better as a `throw new NullReferenceException(nameof(password))`? Since the `Debug.Assert` will compile away, we'll see the `NullReferenceException` thrown on the `password.Length` property access on line 106.

mminns replied:
FWIW this is a straight copy of https://github.com/Microsoft/Git-Credential-Manager-for-Windows/blob/master/GitHub.Authentication/Controls/Text/MaskedPasswordBox.cs

whoisj commented on Bitbucket.Authentication/OAuth/OAuthAuthenticator.cs:
```diff
@@ -0,0 +1,321 @@
+        /// <exception cref="SourceTree.Exceptions.OAuthException">Thrown when OAuth fails for whatever reason</exception>
+        public async Task<AuthenticationResult> GetAuthAsync(TargetUri targetUri, TokenScope scope, CancellationToken cancellationToken)
+        {
+
```
nit: unnecissary vertical space

mminns replied:
done

whoisj commented on Bitbucket.Authentication/OAuth/OAuthAuthenticator.cs:
```diff
@@ -0,0 +1,321 @@
+        /// <returns></returns>
+        private async Task<string> Authorize(TargetUri targetUri, TokenScope scope, CancellationToken cancellationToken)
+        {
+            var authorityUrl = string.Format(
```
nit: this method call is rather difficult to read, any chance you can clean it (and the other below) up?

whoisj commented on Bitbucket.Authentication/OAuth/OAuthAuthenticator.cs:
```diff
@@ -0,0 +1,321 @@
+                    Timeout = TimeSpan.FromMilliseconds(RequestTimeout)
+                })
+                {
+                    var tokenUrl = string.Format(
```
This is a fairly complex call to `String.Format`, I'd rather see string interpolation here.

Example:
```
$"{TokenUri}?grant_type=authorization_code&code={authCode}&client_id={ConsumerKey}&client_secret={ConsumerSecret}&state=authenticated"
```

whoisj commented on Bitbucket.Authentication/OAuth/OAuthAuthenticator.cs:
```diff
@@ -0,0 +1,321 @@
+                                    string responseText = await response.Content.ReadAsStringAsync();
+
+                                    Match tokenMatch;
+                                    if ((tokenMatch = Regex.Match(responseText, @"\s*""access_token""\s*:\s*""([^""]+)""\s*", RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.IgnoreCase)).Success
```
Seems like there's a lot of dynamic regular expressions happen here. I'm guessing that this will have some performance impact. Likely negligible compared to the cost of network operations, but I wanted to point it out.

whoisj commented on Bitbucket.Authentication/OAuth/SimpleServer.cs:
```diff
@@ -0,0 +1,124 @@
+                var context = await listener.GetContextAsync().RunWithCancellation(cancellationToken);
+                rawUrl = context.Request.RawUrl;
+
+                await Task.Delay(100); //Wait 100ms before writing the response out
```
Just for reference, there's no promise made my the CLR that the `Task.Delay(100)` ever gets scheduled, additionally there will be a delay before it is, so the `100` is "just a suggestion" at best.

Why the arbitrary sleep here?

mminns replied:
A good question. We use this code in SourceTree itself and the later delay at line 66 is there to ensure the response HTML page is written out before the server shuts down, but this reason for this one escapes me at the moment, I will check.

whoisj commented on Bitbucket.Authentication/OAuth/SimpleServer.cs:
```diff
@@ -0,0 +1,124 @@
+            {
+                if (!UriParser.IsKnownScheme("pack"))
+                    new System.Windows.Application();
+
```
note: unnecessary vertical space.

mminns replied:
done

whoisj commented on Bitbucket.Authentication/OAuth/SimpleServer.cs:
```diff
@@ -0,0 +1,124 @@
+                    new System.Windows.Application();
+
+
+                var html = Application.GetResourceStream(
```
nit: odd formatting = difficult to read code, please fix.

mminns replied:
done

whoisj commented on Bitbucket.Authentication/OAuth/SimpleServer.cs:
```diff
@@ -0,0 +1,124 @@
+
+                var html = Application.GetResourceStream(
+            new Uri("pack://application:,,,/Bitbucket.Authentication;Component/Assets/auth.html", UriKind.Absolute));
+                if (html != null)
```
Should you validate `html.Stream != null` as well? Otherwise the `new StreamReader(html.Stream)` on line 108 will result in a `ArgumentNullException`.

whoisj commented on Cli-Askpass/Cli-Askpass.csproj:
```diff
@@ -72,8 +72,16 @@
+    </ProjectReference>
     <ProjectReference Include="..\Microsoft.Alm.Authentication\Microsoft.Alm.Authentication.csproj">
-      <Project>{19770407-b493-459d-bb4f-04fbefb1ba13}</Project>
+      <Project>{c1daabc1-b493-459d-bb4f-04fbefb1ba13}</Project>
```
Please do not change the project GUID.

mminns replied:
Hmm not sure how that happened, will undo.

mminns replied:
ah seems to be fixed after the rebase.

whoisj commented on Cli-Shared/Program.cs:
```diff
@@ -747,6 +778,76 @@ private static void EnableTraceLogging(OperationArguments operationArguments, st
+        }
+
+        private static bool BitbucketOAuthPrompt(string title, TargetUri targetUri, Bitbucket.AuthenticationResultType resultType,
+            string username)
```
nit: please either keep all parameters on the same line, or left-align them vertically immediately after the `(` character.

mminns replied:
done

whoisj commented on Cli-Shared/Program.cs:
```diff
@@ -964,14 +1065,15 @@ private static Credential ModalPromptForPassword(TargetUri targetUri, string mes
         private static void PrintVersion()
         {
-            Program.WriteLine($"{Title} version {Version.ToString(3)}");
+            Program.WriteLine($"{Title} version {Version}");
```
Why this change?

mminns replied:
Ah that was during my testing to be able to see #.#.#.# versions from the command line. I'll revert it.

mminns replied:
done

whoisj commented on Cli-Shared/Program.cs:
```diff
@@ -1113,6 +1215,44 @@ private static bool QueryCredentials(OperationArguments operationArguments)
                     break;
 
+                case AuthorityType.Bitbucket:
+					{
```
nit: there's some kind of indentation fiasco happening here - can you please correct? Thanks.

mminns replied:
done

whoisj commented on Cli-Shared/Program.cs:
```diff
@@ -412,6 +424,11 @@ private static Credential BasicCredentialPrompt(TargetUri targetUri, string titl
                             operationArguments.Authority = AuthorityType.GitHub;
                             goto case AuthorityType.GitHub;
                         }
+						else if (authority is Bitbucket.Authentication)
```
nit: white-space issues

mminns replied:
done

whoisj commented on Cli-Shared/Program.cs:
```diff
@@ -1113,6 +1215,44 @@ private static bool QueryCredentials(OperationArguments operationArguments)
+                            LogEvent($"Failed to retrieve Bitbucket credentials for '{operationArguments.TargetUri}'.", EventLogEntryType.FailureAudit);
+                        }
+                    }).Wait();
+					}
```
nit: white-space issues

mminns replied:
Done

whoisj commented:
Also, looks like the change cannot merge and needs to be rebased onto `origin/master`. Thanks!

mminns commented:
Thanks. And no problem about the nits.
I'll work through them and have sweep through the rest of the code and rebase

mminns commented:
I've rebased. Just de-nitting.
I see references in the history to using codemaid to format the code, is that the norm? If so is it just with the out of the box settings?

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+                return targetUri;
+            }
+
+            return
```
done

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+
+                            credentials = GenerateCredentials(targetUri, username, ref result);
+                            SetCredentials(targetUri, credentials, username);
+                            SetCredentials(GetRefreshTokenTargetUri(targetUri),
```
done

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+            BaseSecureStore.ValidateTargetUri(targetUri);
+
+            if (personalAccessTokenStore == null)
+                throw new ArgumentNullException("personalAccessTokenStore",
```
done

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+
+            if (targetUri.ActualUri.DnsSafeHost.EndsWith(BitbucketBaseUrlHost, StringComparison.OrdinalIgnoreCase))
+            {
+                authentication = new Authentication(personalAccessTokenStore
```
done

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+            string password;
+
+            // Ask the user for Basic Auth credentials
+            if (AcquireCredentialsCallback("Please enter your Bitbucket credentials for ", targetUri, out username,
```
done

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+                AuthenticationResult result;
+
+                if (
+                    result =
```
done

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+                if (
+                    result =
+                        await
+                            BitbucketAuthority.AcquireToken(targetUri, username, password,
```
done

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+                    if (AcquireAuthenticationOAuthCallback("", targetUri, result, username))
+                    {
+                        if (
+                            result =
```
done

mminns commented on Bitbucket.Authentication/Authentication.cs:
```diff
@@ -0,0 +1,524 @@
+                    {
+                        if (
+                            result =
+                                await
```
done

mminns commented on Bitbucket.Authentication/Authority.cs:
```diff
@@ -0,0 +1,248 @@
+            get { return "/2.0/user"; }
+        }
+
+        /// <inheritdoc/>
```
force of habit :)

mminns commented on Bitbucket.Authentication/AuthenticationPrompts.cs:
```diff
@@ -0,0 +1,188 @@
+        /// <param name="username">the username entered by the user</param>
+        /// <param name="password"> the password entered by the user</param>
+        /// <returns>returns true if the user provides credentials which are then successfully validated, false otherwise</returns>
+        public static bool CredentialModalPrompt(string title, TargetUri targetUri, out string username,
```
Done

mminns commented on Bitbucket.Authentication/AuthenticationPrompts.cs:
```diff
@@ -0,0 +1,188 @@
+        /// <param name="resultType"></param>
+        /// <param name="username"></param>
+        /// <returns>returns true if the user successfully completes the OAuth dance and the returned access_token is validated, false otherwise</returns>
+        public static bool AuthenticationOAuthModalPrompt(string title, TargetUri targetUri,
```
Done

whoisj commented:
> I see references in the history to using codemaid to format the code, is that the norm? If so is it just with the out of the box settings?

Yes, I do believe it's ITB settings - but I really should check something into the repository. Thanks for the idea, I'll add it to the list. 😄

mminns commented:
I think I pretty much done. Its a long holiday weekend here, but I will do some testing to confirm I've not broken anything and refresh this early next week. Thanks

mminns commented:
Hi, I've rebased, addressed the suggested changes and force pushed my branch. I refactored the OAuthAuthenticator to make it, hopefully, more readable, so I ran through some testing to be happy it still works. All seems good AFAICS.

Thanks for your time looking into this

whoisj commented:
Looks good. There are a few stylistic issues, but I can take care of those in a separate PR.

@mminns thank you very much, and a hearty welcome to you and all of BitBucket to the GCM family! 😄

mminns commented:
Thanks for your help and time. I'll keep an eye on the other PRs, try and pick up the stylistic tips for next time.
